### ISSUE TITLE
Pyright doesn't report type error when passing incorrect generic collection type after narrowing in for-loop

### ISSUE THREAD

**Issue by Nathan-Fenner** [microsoft/pyright#6417]
**Describe the bug**
In the code sample below, Pyright should be reporting a type error, but isn't.

After defining the function

```python
def want(schema: bool):
    # This function should only accept `bool`
    ...
```

and later calling it (incorrectly) as

```python
want(UnrelatedGenericClass(item))
```

we'd expect a type error to appear, since `UnrelatedGenericClass` definitely isn't a `bool`, but Pyright doesn't flag any problem, in the situation detailed below. Attempting to minimize further causes the type-error to appear like it's supposed to.

**Code or Screenshots**

[Pyright 3.11 Playground](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDiApikSEgMYA0UAKokQGoCGIAUG7VALx0MsgAFACJawgJQcKAG2YBnOVACqKEEVkwiAE2KlyFAMKyFg3WUoBtWgF1xALjZQnUYa9oALJIq9RmKXyAARkgwIKxwUACucmQAtFpEwKjaUGgk5hRQMvJyrsKOzgnAUAD6JaghZYIx0sA0FGAoMMyocuWaEHZ09gXOzjXAAHQlDU0tKG0hRNC8o82t7dMcRVAA7n4w1RTu08xdgWBg0j19AMR0norAkSgUMEiNUHLuYJHSWlCN0hHMFBRECHwAAMDkcgb0oIMoctElAiAAPZjYaREErXW6CCFySCohrSFF3B4oLrSLwwCyg6TWKgQqadTBNKAAHyeoRpJ2c5w8RCgQOAuCBUGkhwQvhQHyBSGAgrkzQ6JHwrB5Bxg7igagAjpEkGoPvyQLBLlBApEMDAwL4EAgiKwuhD9aUGU8cSMjgT7o07X0%2BlLMB0nTFNhYAIw0ABMNAAzLYHN642sNoIVGoNNozPojDlBHTxOIoOcADyxWJQAAKcHIaHc%2BC0YCIchQAHJ8A1keNHGxzusmknVOpmJodOkM8Y5IIAKy5-NQIsl303VsQBXaGiqnyk0hQdzyXywBhwkDgEA0UkAax5IUbijkkStYBiH3NQA)

[Pyright 3.12 playground](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=MYGwhgzhAECqB2AnApuALsgJgcWfZiAlsAMLhQDaAKgLoBcAUNM9AETtUAWhMP0Y8fogBGhNIjCIAntACuEAgFpMyAGaF8maAHM8BYtFCQI7VkxYrV0APrWNY2wAoFIVQBpDAe3howGiHYYALZ00FQAlIws0dAuqgB01sDevv6ByEHQALxePn7wAWIZDAyW0ADuAmjOwJwZYKHCnp4gkebMAMRh3DCqsvDAaITesZyesiBa3iAyYMDAyAAOaNAABk0tq%2B3Q8bslZcgAHmBBiyDI1n0DjtsQnkEXySDng8PwoSA8aBQbIDRu2yKIWgGhWAB9YuIAW1ol0uMg1qpPIhVtAQM1Fvx4FpVoRVKiIL5gngVpIEU00JxoCgAI6yQgoLRIxDQSl8YSybSszz8RaLZCSULbZk2EGCO4PJItF5DbxCmIsPEg4Ji2LIaoUACMHgATB4AMw0GEKmKVHyOBAodBYXD4IikcgQRxA8LhaBdAA8ikU0AAClIiNpOCtMJ5kBB4AByFbJU7gDRMBhdM3VS2oMAYHB6e1kYyOACsrvd0C9PqV-VjDx8WA8bJgn3w0E4kH4rKk-OgBEQyI8nwA1gixJGYBBZHzPAotGhPEA)

```python
# Python 3.11
from typing import Generic, TypeVar

T = TypeVar("T")

class UnrelatedGenericClass(Generic[T]):
    """This is an arbitrary user-defined generic class"""
    def __init__(self, contains_item: T):
        self._contains_item = contains_item

def want(schema: bool):
    # This function should only accept `bool`
    ...

def example_func(
    some_collection: list[bool],
    item: int | str,
):
    # The `for` loop and `if` statement are both required for this bug to appear: 
    for _ in some_collection: 
        if item in set([1, 2, 3]):
            want(UnrelatedGenericClass(item)) # <-- Pyright doesn't complain
 
# want(UnrelatedGenericClass(5)) # <-- if uncommented, this line has a type error, like it's supposed to
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
- I first encountered this in VS Code Pylance v2023.11.10
- It reproduces in the playground link above as version 1.1.335

---

While I was minimizing this bug, I noticed that both VS Code Pylance and the Pyright Playground show some non-determinism relating to this problem. Sometimes this line _will_ be reported as an error, but then the error goes away when I begin typing in the file.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.339, which I just published. It will also be included in a future release of pylance.
